materialize-boilerplate/apply.go

```diff
@@ -0,0 +1,229 @@
+// optionally with a concurrent scatter/gather for expedience on endpoints that would benefit from
+// that sort of thing.
+func ApplyChanges(ctx context.Context, req *pm.Request_Apply, applier Applier, is *InfoSchema, concurrent bool) (*pm.Response_Applied, error) {
+	if err := req.Validate(); err != nil {
+		return nil, fmt.Errorf("validating request: %w", err)
+	}
+
+	storedSpec, err := applier.LoadSpec(ctx, req.Materialization.Name)
+	if err != nil {
+		return nil, fmt.Errorf("getting stored spec: %w", err)
+	}
+
+	actionDescriptions := []string{}
+	actions := []ActionApplyFn{}
+
+	addAction := func(desc string, a ActionApplyFn) {
+		if a != nil { // Convenience for handling endpoints that return `nil` for a no-op action.
+			actionDescriptions = append(actionDescriptions, desc)
+			actions = append(actions, a)
+		}
+	}
+
+	// Only attempt to create meta tables if the storedSpec was `nil`, which suggests that the meta
+	// tables don't yet exist.
+	// TODO(whb): We will eventually stop persisting specs for materializations, and instead include
+	// the previous spec as part of the protocol. When that happens this will need updated.
+	if storedSpec == nil {
+		desc, action, err := applier.CreateMetaTables(ctx, req.Materialization)
+		if err != nil {
+			return nil, fmt.Errorf("getting CreateMetaTables action: %w", err)
+		}
+		addAction(desc, action)
+	}
+
+	for bindingIdx, binding := range req.Materialization.Bindings {
+		// The existing binding spec is used to extract various properties that can't be learned
+		// from introspecting the destination system, such as the backfill counter and if the
+		// materialization was previously delta updates.
+		existingBinding, err := FindExistingBinding(binding.ResourcePath, binding.Collection.Name, storedSpec)
+		if err != nil {
+			return nil, fmt.Errorf("finding existing binding: %w", err)
+		}
+
+		if !is.HasResource(binding.ResourcePath) {
+			// Resource does not yet exist, and must be created.
+			desc, action, err := applier.CreateResource(ctx, req.Materialization, bindingIdx)
+			if err != nil {
+				return nil, fmt.Errorf("getting CreateResource action: %w", err)
+			}
+			addAction(desc, action)
+		} else if existingBinding != nil && existingBinding.Backfill != binding.Backfill {
+			// Resource does exist but the backfill counter is being increased, so it must be
+			// replaced.
+			desc, action, err := applier.ReplaceResource(ctx, req.Materialization, bindingIdx)
+			if err != nil {
+				return nil, fmt.Errorf("getting ReplaceResource action: %w", err)
+			}
+			addAction(desc, action)
+		} else {
+			// Resource does exist and may need updated for changes in the binding specification.
+			params := BindingUpdate{
+				NewlyDeltaUpdates: existingBinding != nil && !existingBinding.DeltaUpdates && binding.DeltaUpdates,
+			}
+
+			for _, proposedField := range binding.FieldSelection.AllFields() {
```

Comment:
nit: you could shorten up the line length by proposedField => field and proposedProjection => projection without loss of clarify IMO.